src/compiler/scala/tools/nsc/transform/LazyVals.scala

```diff
@@ -100,11 +115,26 @@ abstract class LazyVals extends Transform with TypingTransformers with ast.TreeD
             val (rhs1, sDef) = mkLazyDef(enclosingClassOrDummyOrMethod, transform(rhs), idx, sym)
             sym.resetFlag((if (lazyUnit(sym)) 0 else LAZY) | ACCESSOR)
             (rhs1, sDef)
-          } else
+          } else if (sym.hasAllFlags(MODULE | METHOD) && !sym.owner.isTrait) {
+            rhs match {
+              case b @ Block((assign @ Assign(moduleRef, _)) :: Nil, expr) =>
+                val cond = Apply(Select(moduleRef, Object_eq), List(Literal(Constant(null))))
+                val (fastPath, slowPath) = mkFastPathBody(sym.owner.enclClass, moduleRef.symbol, cond, transform(assign) :: Nil, Nil, transform(expr))
```

Comment:
ping @retronym -- I think the sharing of `cond` is a problem, the rest of the PR LGTM